File: evd_ros_backend/evd_ros_core/src/evd_script/program/primitives/delay.py
from ..primitive import Primitive


class Delay(Primitive):

    '''
    Data structure methods
    '''
    def __init__(self, duration=0, type='', name='', uuid=None, parent=None, append_type=True):
        self._duration = None

        super(Delay,self).__init__(
            type='delay.'+type if append_type else type,
            name=name,
            uuid=uuid,
            parent=parent,
            append_type=append_type)

        self.duration = duration

    def to_dct(self):
        msg = super(Delay,self).to_dct()
        msg.update({
            'duration': self.duration,
        })
        return msg

    @classmethod
    def from_dct(cls, dct):
        return cls(
            name=dct['name'],
            type=dct['type'],
            append_type=False,
            uuid=dct['uuid'],
            duration=dct['duration'])

    '''
    Data accessor/modifier methods
    '''

    @property
    def duration(self):
        return self._duration

    @duration.setter
    def duration(self, value):
        if self._duration != value:
            self._duration = value
            self.updated_attribute("duration",'set')

    def set(self, dct):
        duration = dct.get('duration', None)
        if duration != None:
            self.duration = duration

        super(Delay,self).set(dct)

    '''
    Update Methods
    '''

    def deep_update(self):

        super(Delay,self).deep_update()

        self.updated_attribute('duration','update')

    def shallow_update(self):
        super(Delay,self).shallow_update()

        self.updated_attribute('duration','update')

File: evd_authoring_app/src/model/evdScript/program/primitives/delay.js
import { Primitive } from '../primitive';
import Blockly from 'blockly';

export class Delay extends Primitive {

    /*
    * Data structure methods
    */

    static typeString() {
        return 'delay.';
    }

    static fullTypeString() {
        return Primitive.fullTypeString() + Delay.typeString();
    }

    constructor(duration=0, type='', name='', uuid=null, parent=null, appendType=true) {
        super(
            (appendType) ? 'delay.'+type : type,
            name,
            uuid,
            parent,
            appendType
        );

        this._duration = null;

        this.duration = duration;
    }

    toDict() {
        const msg = {
            ...super.toDict(),
            duration: this.duration
        };
        return msg;
    }

    static fromDict(dct) {
        return new Delay(
            dct.duration,
            dct.type,
            dct.name,
            dct.uuid,
            null,
            false
        );
    }

    static BlocklyToolbox() {
        return { type: 'delay' };
    }

    static BlocklyBlock() {
        return { key: 'delay', data: {
            init: function() {
                this.appendDummyInput()
                    .appendField("Delay");
                this.appendDummyInput()
                    .appendField("Duration")
                    .appendField(new Blockly.FieldNumber(1, 0), "duration")
                    .appendField("seconds");
                this.setPreviousStatement(true, null);
                this.setNextStatement(true, null);
                this.setColour(120);
                this.setTooltip("delay");
                this.setHelpUrl("delay");
            }
        }};
    }

    /*
     * Data accessor/modifer methods 
     */

     get duration() {
         return this._duration;
     }

     set duration(value) {
         if (this._duration !== value) {
             this._duration = value;
         }

         super.updatedAttribute('duration','set');
     }

     set(dct) {
         if ('duration' in dct) {
             this.duration = dct.duration;
         }

         super.set(dct);
     }

     /*
     * Update methods
     */

     deepUpdate() {
         super.deepUpdate();

         this.updatedAttribute('duration','update');
     }

     shallowUpdate() {
         super.shallowUpdate();

         this.updatedAttribute('duartion','update');
     }
}